Index: drivedb.h
===================================================================
--- drivedb.h	(revision 3156)
+++ drivedb.h	(working copy)
@@ -238,7 +238,7 @@
     "(IBM-|Hitachi )?IC35L0[12346]0AVER07.*",
     "",
     "IBM Deskstar 60GXP drives may need upgraded SMART firmware.\n"
-    "Please see http://www.geocities.com/dtla_update/index.html#rel and\n"
+    "Please see http://haque.net/dtla_update/ and\n"
     "http://www.ibm.com/pc/support/site.wss/MIGR-42215.html",
     ""
   },
@@ -251,7 +251,7 @@
     "(IBM-)?DTLA-30[57]0[123467][05].*",
     "",
     "IBM Deskstar 40GV and 75GXP drives may need upgraded SMART firmware.\n"
-    "Please see http://www.geocities.com/dtla_update/ and\n"
+    "Please see http://haque.net/dtla_update/ and\n"
     "http://www.ibm.com/pc/support/site.wss/MIGR-42215.html",
     ""
   },
