Index: Makefile.am
===================================================================
--- Makefile.am	(revision 3473)
+++ Makefile.am	(working copy)
@@ -544,7 +544,8 @@
           -e '/^\.\\" %IF OS .*$(os_man_filter)/,/^\.\\" %ENDIF OS .*$(os_man_filter)/ s,^,!!,' \
           -e '/^\.\\" %IF OS ./,/^\.\\" %ENDIF OS ./ s,^,.\\"\# ,' \
           -e '/^!*\.\\" %IF NOT OS ./,/^!*\.\\" %ENDIF NOT OS ./ s,^,!!,' \
-          -e '/^!!/{ s,^!!!*,,; s,^\.\\"! \(.*\)$$,\1 \\"\#, ;}' ; \
+          -e 's,^!!!*\.\\"! \(.*\)$$,\1 \\"\#,' \
+          -e 's,^!!!*,,' ; \
     else \
       cat; \
     fi
