--- drivedb.h.orig	2020-07-15 12:42:43.408397412 -0300
+++ drivedb.h	2020-07-15 13:39:06.426187582 -0300
@@ -2043,6 +2043,7 @@
     "Patriot P200 ((128|256|512)GB|[12]TB)|" // tested with Patriot P200 256GB/S1022A0
     "R3SL(120|240|480|960)G|" // AMD Radeon SSDs, tested with R3SL240G/P0422C
     "Ramsta SSD S800 (120|240|480)GB|" // SM2258G, tested with Ramsta SSD S800 480GB/RS81V0
+    "SATA3 ((120|240|480)G|(1|2)T)B SSD|" // TCSUNBOW X3, tested with SATA3 1TB SSD/S1230A0, SATA3 240GB SSD/S0618A0 
     "T60|" // KingSpec T60, tested with T60/20151120
     "TCSUNBOW [MX]3 (60|120|240)GB|" // TC-Sunbow M3/X3, tested with TCSUNBOW M3 240GB/R0327B0,
        // TCSUNBOW X3 120GB/R1211A0
