--- drivedb.h	2022-04-28 00:31:40.000000000 +0300
+++ drivedb.h.new	2022-04-30 20:52:59.060261751 +0300
@@ -3848,6 +3848,14 @@
     "TOSHIBA HDWD2[246]0",
     "", "", ""
   },
+  { "Toshiba S300 (CMR)",
+    "TOSHIBA HDW(T140|V110|T3(60|80|1A))",
+    "", "", ""
+  },
+  { "Toshiba S300 (SMR)", // tested with TOSHIBA HDWT860/KQ0H1L
+    "TOSHIBA HDWT(7[24]|8[46])0",
+    "", "", ""
+  },
   { "Toshiba X300", // tested with TOSHIBA HDWE160/FS2A, TOSHIBA HDWF180/GX0B
     "TOSHIBA HDW(E1[456]0|[FR]180|R(11A|21[CE]|31G))",  // 11A:10TB, 21C:12TB, 21E:14TB, 31G: 16TB
     "", "",
