Index: ataprint.cpp
===================================================================
--- ataprint.cpp	(revision 4602)
+++ ataprint.cpp	(working copy)
@@ -3687,7 +3687,7 @@
     bool use_gplog = true;
     unsigned nsectors = 0;
     if (gplogdir) 
-      nsectors = GetNumLogSectors(gplogdir, 0x04, false);
+      nsectors = GetNumLogSectors(gplogdir, 0x04, true);
     else if (smartlogdir){ // for systems without ATA_READ_LOG_EXT
       nsectors = GetNumLogSectors(smartlogdir, 0x04, false);
       use_gplog = false;
