--- os_freebsd.cpp	2010-01-28 21:24:32.000000000 +0100
+++ os_freebsd.cpp	2010-10-10 18:35:38.000000000 +0200
@@ -317,6 +317,9 @@
 
   // ata_28bit_cmd
   ccb.ataio.cmd.flags = 0;
+  if (request->flags & ATA_CMD_CONTROL) {
+    ccb.ataio.cmd.flags |= CAM_ATAIO_NEEDRESULT;
+  }
   ccb.ataio.cmd.command = request->u.ata.command;
   ccb.ataio.cmd.features = request->u.ata.feature;
   ccb.ataio.cmd.lba_low = request->u.ata.lba;
@@ -331,7 +334,9 @@
     err(1, "cam_send_ccb");
     return -1;
   }
-
+  
+  request->u.ata.count = ccb.ataio.res.sector_count;
+  
   if ((ccb.ccb_h.status & CAM_STATUS_MASK) == CAM_REQ_CMP)
     return 0;
 
