*** drivedb.h.orig	2018-08-19 11:45:53.000000000 -0700
--- drivedb.h	2018-08-21 13:56:36.000000000 -0700
***************
*** 304,310 ****
      "Micron_M500DC_(EE|MT)FDDA[AK](120|240|480|800)MBB|" // tested with Micron_M500DC_EEFDDAA120MBB/129, Micron_M500DC_MTFDDAK800MBB/0129
      "Micron_1100_MTFDDA[KV](256|512|1T0|2T0)TBN|" // Marvell 88SS1074, tested Micron_1100_MTFDDAK256TBN/M0MU020
      "MICRON_M510DC_(EE|MT)FDDAK(120|240|480|800|960)MBP|" // Micron_M510DC_MTFDDAK240MBP/0005
-     "Micron_5100_(EE|MT)FDDA[KV](240|480|960|1T9|3T8|7T6)T(BY|CB|CC)", // tested with Micron_5100_MTFDDAK1T9TBY
      "", "",
    //"-v 1,raw48,Raw_Read_Error_Rate "
      "-v 5,raw48,Reallocate_NAND_Blk_Cnt "
--- 304,309 ----
***************
*** 332,337 ****
--- 331,370 ----
      "-v 247,raw48,Host_Program_Page_Count "
      "-v 248,raw48,Bckgnd_Program_Page_Cnt"
    },
+ // Reference: https://www.micron.com/resource-details/feec878a-265e-49a7-8086-15137c5f9011
+ // TN-FD-34: 5100 SSD SMART Implementation
+   {
+     "Micron 5100 Pro / 5200 drives",
+     "(Micron_5100_)?(EE|MT)FDDA[KV](240|480|960|1T9|3T8|7T6)T(BY|CB|CC)|" // Matches both stock and Dell OEM
+     "(Micron_5200_)?MTFDDAK(480|960|1T9|3T8|7T6)TD(C|D|N)",
+     "", "",
+     "-v 1,raw48,Raw_Read_Error_Rate "
+     "-v 5,raw48,Reallocated_Block_Count "
+ //  "-v 9,raw24(raw8),Power_On_Hours "  // raw24(raw8)??
+     "-v 12,raw48,Power_Cycle_Count "
+     "-v 170,raw48,Reserved_Block_Pct " // Percentage of remaining reserved blocks available
+     "-v 171,raw48,Program_Fail_Count "
+     "-v 172,raw48,Erase_Fail_Count "
+     "-v 173,raw48,Avg_Block-Erase_Count "
+     "-v 174,raw48,Unexpect_Power_Loss_Ct "
+     "-v 180,raw48,Reserved_Block_Count " // absolute count of remaining reserved blocks available
+     "-v 183,raw48,Error_Correction_Count " // SATA speed downshift count
+     "-v 184,raw48,Error_Correction_Count "
+     "-v 187,raw48,Reported_Uncorrect " // Number of UECC correction failures
+     "-v 188,raw48,Command_Timeouts "
+     "-v 194,tempminmax,Temperature_Celsius " // 100 - degrees C, wraps: 101 reported as 255, 102 as 254 etc. May be best to use the raw value as it's unclear how tempminmax is interpreted
+     "-v 195,raw48,Cumulativ_Corrected_ECC "
+     "-v 196,raw48,Reallocation_Event_Ct "
+     "-v 197,raw48,Current_Pending_Sector " // Use the raw value
+     "-v 198,raw48,Offline_Uncorrectable "  // Use the raw value
+     "-v 199,raw48,UDMA_CRC_Error_Count "   // Use the raw value
+     "-v 202,raw48,Percent_Lifetime_Remain " // Remaining endurance, trips at 10%
+     "-v 206,raw48,Write_Error_Rate "
+     "-v 210,raw48,Write_Error_Rate "       // Total number of NAND pages recovered by RAIN
+     "-v 247,raw48,Host_Program_Page_Count "
+     "-v 248,raw48,Bckgnd_Program_Page_Cnt"
+ 
+   },
    { "Micron M500DC/M510DC Enterprise SSDs",
      "Micron_M500DC_(EE|MT)FDDA[AK](120|240|480|800)MBB|" // tested with
        // Micron_M500DC_EEFDDAA120MBB/129, Micron_M500DC_MTFDDAK800MBB/0129
